### DISCUSSION TITLE
posts/Go/pointer/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#352] [Announcements]
https://www.liwenzhou.com/posts/Go/pointer/

**Comment by chengkanghua** [comment]
在最新版本的go上实验 输出的内存地址都不一样
```

PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
package main

import "fmt"

func main() {
        a := 10
        b := &a
        fmt.Printf("a:%d prt:%d \n", a, &a)
        fmt.Printf("b:%p type:%T \n", b, b)
        fmt.Println(&b)
}
PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
a:10 prt:824634130520
b:0xc000064058 type:*int
0xc000090018
```

**Comment by Q1mi** [comment]
> @chengkanghua
> 在最新版本的go上实验 输出的内存地址都不一样
> ```
> 
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> func main() {
>         a := 10
>         b := &a
>         fmt.Printf("a:%d prt:%d \n", a, &a)
>         fmt.Printf("b:%p type:%T \n", b, b)
>         fmt.Println(&b)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> a:10 prt:824634130520
> b:0xc000064058 type:*int
> 0xc000090018
> ```
注意搞清楚变量存的值和变量的内存地址的区别：
func main() {
	a := 10
	b := &a
	fmt.Printf("a:%d prt:%v \n", a, &a)
	// b保存的值是a的内存地址
	fmt.Printf("b:%v type:%T \n", b, b)
	fmt.Println(&b) // b自己的内存地址
}

**Comment by code4happylife** [comment]
```go
package main
import "fmt"
func main() {
	a := 10
	b := &a
	fmt.Printf("a:%d ptr:%p\n", a, &a)//a:10 ptr:0xc000066080
	fmt.Printf("b:%p type:%T\n", b, b)//b:0xc000066080 type:*int
	fmt.Println(&b)//0xc000092018

	c := 12.0
	d := &c
	fmt.Printf("c : %v, &c : %p ,type of c: %T\n",c,&c,c)//c : 12, &c : 0xc000066098 ,type of c: float64
	fmt.Printf("d: %v, &d: %p, *d:%v,type of d:%T\n",d,&d,*d,d)//d: 0xc000066098, &d: 0xc000092028, *d:12,type of d:*float64


	var e map[string]int
	fmt.Println(e == nil)//true
	fmt.Printf("type of e: %T\n",e)//type of e: map[string]int
	e = make(map[string] int,1)
	e["score"] = 100
	fmt.Println(e)//map[score:100]

	var f *int
	fmt.Println(f == nil)//true
	f = new(int)
	*f = 100
	fmt.Printf("value of f pointed :%v",*f)//value of f pointed :100

}

```

**Comment by code4happylife** [comment]
> @chengkanghua
> 在最新版本的go上实验 输出的内存地址都不一样
> ```
> 
> PS C:\Users\VULCAN\go_learm\src\ch2\test> cat .\test.go
> package main
> 
> import "fmt"
> 
> func main() {
>         a := 10
>         b := &a
>         fmt.Printf("a:%d prt:%d \n", a, &a)
>         fmt.Printf("b:%p type:%T \n", b, b)
>         fmt.Println(&b)
> }
> PS C:\Users\VULCAN\go_learm\src\ch2\test> go run .\test.go
> a:10 prt:824634130520
> b:0xc000064058 type:*int
> 0xc000090018
> ```

```go
package main

import "fmt"

func main() {
	a := 10
	b := &a
	fmt.Printf("a:%d prt:%p \n", a, &a)
	fmt.Printf("a:%d prt:%d \n", a, &a)
    fmt.Printf("the value of %d represented in Hex is: %x\n",&a,&a)
	fmt.Printf("b:%p type:%T \n", b, b)
	fmt.Println(&b)
}
/*a:10 prt:0xc0000120b8
a:10 prt:824633794744
the value of 824633794744 represented in Hex is: c0000120b8
b:0xc0000120b8 type:*int
0xc000006028*/```

**Comment by 01super** [comment]
up！ up！ up！

**Comment by shuitianyise** [comment]
这个指针讲的我感觉 内容有点少 楼主是不是充实一下 比如 指针做函数返回值 指针做参数

**Comment by SmeagolYe** [comment]
done

**Comment by choyda** [comment]
> @shuitianyise
> 这个指针讲的我感觉 内容有点少 楼主是不是充实一下 比如 指针做函数返回值 指针做参数

在函数的时候会讲
